### PR TITLE
Display fast/slow alongside judges

### PR DESCRIPTION
**Pull request by Tina-otoge**
This PR adds a "Fast" or "Slow" text above judges when not hitting perfectly.

It integrates well in osu and mania rulesets, looks a bit weird in taiko, not completely sure if I could take a different approach for taiko.

Also I should probably need to add test cases for this?

Maybe this should also be made into a toggle in user options near hit error bar settings?

### CODE REVIEW

File: osu.Game/Rulesets/Judgements/JudgementResult.cs [hunk 1]
```diff
@@ -30,6 +30,8 @@ public class JudgementResult
         [NotNull]
         public readonly Judgement Judgement;
 
+        public HitDetail Detail => TimeOffset < 0 ? HitDetail.Early : HitDetail.Late;
```
Comment (iiSaLMaN) [inline osu.Game/Rulesets/Judgements/JudgementResult.cs]: What happens when hitting at perfect timing `TimeOffset = 0`?

File: osu.Game/Rulesets/Scoring/HitResult.cs [hunk 2]
```diff
@@ -44,4 +44,13 @@ public enum HitResult
...
         Perfect,
     }
+
+    public enum HitDetail
+    {
+        [Description(@"Fast")]
+        Early,
+
+        [Description(@"Slow")]
+        Late,
```
Comment (iiSaLMaN) [inline osu.Game/Rulesets/Scoring/HitResult.cs]: Pretty much unnecessary to add descriptions here, just name the enums `Fast` and `Late` then use `HitDetail.ToString()`.

**Review by frenzibyte — CHANGES_REQUESTED** [review]

File: osu.Game/Rulesets/Scoring/HitResult.cs [hunk 3]
```diff
@@ -44,4 +44,13 @@ public enum HitResult
         [Description(@"Perfect")]
         Perfect,
     }
+
+    public enum HitDetail
```
Comment (iiSaLMaN) [inline osu.Game/Rulesets/Scoring/HitResult.cs]: Should have it's own `.cs` file.

File: osu.Game/Rulesets/Judgements/JudgementResult.cs (hunk 1 above)
Comment (iiSaLMaN) [inline osu.Game/Rulesets/Judgements/JudgementResult.cs]: xmldoc.

Comment (iiSaLMaN) [reply osu.Game/Rulesets/Judgements/JudgementResult.cs]: Just realized that no detail text appears on `Great` judgement, although I advocate for adding a `Perfect` detail rather than having a false value.

Comment (skielred) [reply osu.Game/Rulesets/Judgements/JudgementResult.cs]: It's hidden when `TimeOffset` is within the window of a Perfect or Great judgement, because the judgement text already gives the information that we hit perfectly.

Comment (skielred) [reply osu.Game/Rulesets/Judgements/JudgementResult.cs]: Alright, actually done that at first but could not find a use-case where it would be used. If you think it's more fitting to have a third member in the Enum I'll be adding one

File: osu.Game/Rulesets/Judgements/DrawableJudgement.cs [hunk 4]
```diff
@@ -134,5 +146,32 @@ private Color4 judgementColour(HitResult judgement)
...
+        {
+            switch (result.Type)
+            {
+                case HitResult.Perfect:
+                case HitResult.Great:
+                    return "";
+                default:
+                    return result.Detail.GetDescription().ToUpperInvariant();
+            }
+        }
```
Comment (iiSaLMaN) [inline osu.Game/Rulesets/Judgements/DrawableJudgement.cs]: What about `HitResult.Miss`?

**Comment by iiSaLMaN** [comment]
Seems OK visually, the colours of `Fast` and `Slow` seems to be not fitting with the judgment result. How about Green for `Fast` and Yellow or Light Red for `Slow`?

**Comment by skielred** [comment]
Blue for `Fast` and red for `Slow` are kinda standard across all rhythm games, sometime text is different, but the colors are blue and red in every rhythm game I can think of that has this feature. Also the contrast helps hinting your timing without actually having to read the text.
Haven't really played with other colors, I can try Light Red for `Slow`, but I believe we should stick to blue for `Fast`, maybe other hues of blue blends in better?

File: osu.Game/Rulesets/Judgements/DrawableJudgement.cs [hunk 5]
```diff
@@ -65,13 +66,24 @@ private void load(OsuColour colours)
...
+                        Text = Result.Type.GetDescription().ToUpperInvariant(),
+                        Font = OsuFont.Numeric.With(size: 20),
+                        Colour = judgementColour(Result.Type),
+                        Scale = new Vector2(0.85f, 1),
+                    }, confineMode: ConfineMode.NoScaling),
+                    new SkinnableDrawable(new GameplaySkinComponent<HitDetail>(Result.Detail), _ => JudgementDetailText = new OsuSpriteText
+                    {
+                        Text = detailText(Result),
+                        Font = OsuFont.Numeric.With(size: 15),
+                        Y = -20,
```
Comment (skielred) [inline osu.Game/Rulesets/Judgements/DrawableJudgement.cs]: This -20 is kinda arbitrary, I may be missing a way to display this component above the other one nicely

**Comment by iiSaLMaN** [comment]
> Haven't really played with other colors, I can try Light Red for `Slow`, but I believe we should stick to blue for `Fast`, maybe other hues of blue blends in better?

how about Cyan for `Fast`, fits well with light red I believe.

**Comment by skielred** [comment]
Cyan and RedLight looks really nice! Thank you for the heads up

File: osu.Game/Rulesets/Scoring/HitDetail.cs [hunk 6]
```diff
@@ -0,0 +1,23 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+namespace osu.Game.Rulesets.Scoring
+{
+    public enum HitDetail
```
Comment (swoolcock) [inline osu.Game/Rulesets/Scoring/HitDetail.cs]: I don’t think this really warrants its own enum, it could just be done directly from the `TimeOffset`.

Either way, `HitResultAccuracy` is a better name, “Detail” is very arbitrary.

File: osu.Game/Rulesets/Judgements/DrawableJudgement.cs (hunk 5 above)
Comment (iiSaLMaN) [reply osu.Game/Rulesets/Judgements/DrawableJudgement.cs]: you're probably looking for a `FillFlowContainer`, check its usages for understanding on how to use it.

**Review by swoolcock — COMMENTED** [review]
Since this is a feature that affects gameplay, this PR needs:

* An accompanying proposal request issue
* Designs from the designer

File: osu.Game/Rulesets/Judgements/DrawableJudgement.cs (hunk 5 above)
Comment (VPeruS) [reply osu.Game/Rulesets/Judgements/DrawableJudgement.cs]: FlowContainer with Direction = FillDirection.Vertical

**Comment by GSculerlor** [comment]
The indicator overlapping the skinnable hit result https://streamable.com/sgyz6

**Comment by skielred** [comment]
@GSculerlor Yeah that was the purpose of my question in https://github.com/ppy/osu/pull/7310#discussion_r360428851
I'll try using `FlowContainer` this should fix the overlapping issue

**Comment by Weigsu** [comment]
### Problem
I feel this creates too much visual noise. Specially as the default skin is quite overwhelming already.
Also, if I understood correctly of how the new skins work I will be able to skin the HitResult (perfect, good, meh) and HitDetail (fast, slow) separately. This however doesn't offer enough flexibility for skinning in my opinion.

### My solution
My preferred way to skin this would be to leave only "Good", "Meh" and "Miss" texts. (and nothing on slider ends) and then color the text yellow if I'm going too fast and blue if too slow.
[Example of that (left is unchanged, right is as described above)](https://streamable.com/dxoxg)

### Possible implementation?
This can't be done with current implementation where I have to skin HitResult and HitDetail separately. And seems there is no standard way to have GameplaySkinComponent based on two values. I'm not quite sure how to do it.
Maybe combine HitResult and HitDetail in struct and implement ToString like that:
``` C#
struct MyHitStruct
{
    private HitResult HitResult;
    private HitDetail HitDetail;

    public MyHitStruct(HitResult hitResult, HitDetail hitDetail)
    {
        HitResult = hitResult;
        HitDetail = hitDetail;
    }

    public override string ToString()
    {
        return string.Join("/", HitResult.ToString(), HitDetail.ToString());
    }
}
// and then when used
new SkinnableDrawable(new GameplaySkinComponent<MyHitStruct>(new MyHitStruct(Result.Type, Detail)), ....)
```
or just add all possible combinations to HitResult enum, but this seems even worse.

### Motivation
Having the extra text was too much to process for me while playing and I just ignored it.  But when I color coated the HitResults it actually helped me to get into correct rhythm sometimes (not as much as I hoped though)

But the best effect this had was that I didn't feel the constant need to look at the Hit Error bars to see if I clicked too late or early (and then missing or losing acc because of that). This made the game experience feel better and "smoother/less abrasive?" for me.

Love this feature and I hope it will make it to lazer!

**Comment by bdach** [comment]
@Weigzu Your proposal (or something extremely similar to it) has already been submitted as a PR once and rejected; see #3210 for rationale.

**Comment by Weigsu** [comment]
> @Weigsu Your proposal (or something extremely similar to it) has already been submitted as a PR once and rejected; see #3210 for rationale.

This is very sad. I don't quite understand the reason for rejection though, peppy said in proposal that he liked the idea (so you're saying there is a chance? :) but in PR just that the colors shouldn't be applied directly to HitResult? Not a big issue

Just to be clear, the color thing I made example of was just one use case (of how I would use it) but the default skin would remain unchanged.
There could be skinnable object for every HitResult HitDetail combo like so:
 * `HitResult` - if the skin doesn't support fast/slow displaying or when it is disabled from options
 * `HitResult/HitDetail.Fast` - for every type of HitResult a skinnable object for when click was too early
 * `HitResult/HitDetail.Slow` - ... was too late
 * `HitResult/HitDetail.Flawles` - ...was on time and some HitResult Judgements are always flawless

Right now (in this PR) there are skinnable objects just for every HitResult and for every HitDetail but not for combinations of them.

I don't think this feature would be good addition to default skin, but it would be awesome to be able to apply custom skin elements to them  (if the skin author desires to).

**Comment by bdach** [comment]
New legacy skin elements are also unlikely to happen ([see comment](https://github.com/ppy/osu/issues/3209#issuecomment-533942709)).

**Comment by skielred** [comment]
Coloring the judge or adding a prefix or suffix (like + or -) are both options that were also discussed when I first talked about implementing this alongside my friends, and I think it's a good compromise for people who think it adds too much elements on screen and makes reading too complex.
It is already noticeable that people are bothered when there are too much elements on screen as most people hide Perfect/Greats from gameplay, most people want the less visual clutter possible.

However, adding an additional elements also allow compatibility with legacy skins by not modifying the behavior of the current judges element, while making it possible to make new skins that hide the judges and instead only display FAST or SLOW, both, or a "-" and "+" on the side of judges instead of "FAST" and "SLOW" above, etc.

As discussed before, I (or someone else) will have to submit a proposal request issue and discuss what works best for osu! in it.

My personal opinion is that it should be left as an option to the player or the skinner to decide how the "detail" / additional accuracy info should be displayed.

Either way, I think this is an important feature as reading the Hit Error Display bar is not intuitive at all while being aware of if we are hitting too early or too late is very much needed to adjust our typing while playing.

I'm completely willing to work again on this PR when we'll have discussed what would be the most fitting implementation for osu!.

**Comment by Weigsu** [comment]
> New legacy skin elements are also unlikely to happen ([see comment](https://github.com/ppy/osu/issues/3209#issuecomment-533942709)).

I don't see any reason to add new legacy skin elements.

> Either way, I think this is an important feature as reading the Hit Error Display bar is not intuitive at all...

Yes, exactly. And it's hard to look at sides of the screen and keep playing :)

#### Clarification
I'll try to explain my issue with current implementation again.
If I wanted to hide "Great" judgments but show custom slow/fast elements on other judgments ("Meh", "Good") it would still display the slow/fast over "Great" even though "Great" itself is invisible. Also it would display it over misses etc.
Demonstrated in video in row "PR issue"

I just feel that if adding extra elements to Judgment  is the only way to skin, it won't be used very much. I made a video with two examples that I believe would be usable for gameplay. But what I think is often very different to what everyone else thinks. If it is so I can just play in dev mode no problem.

[Video with examples.](https://streamable.com/bnofn)

The two examples:
 * Color - Judgment text is colored based on fast/slow (yellow fast, blue slow in this case)
 * Text - Color is kept default but the text is changed to fast/slow (or +/- etc) Also the "Great" judgment would be hidden but it's there to show default colors.

#### In conclusion:
 * No new legacy skin elements added
 * Not supported in old osu, only in lazer
 * Optional for skinner to add
 * If skin is missing HitDetail (fast/slow) elements then this option will be disabled/hidden and HitResult skin is used
 * Not part of default lazer skin (unless someone figures out a way this could fit in default skin)
 
Also, here's a [osu community feature request](https://osu.ppy.sh/community/forums/topics/167932). Might be useful.

**Comment by smoogipoo** [comment]
Personally quite like the "text" version in that example video. It could be adapted very well to more minimalistic skins that use dots - I imagine skins having up/down arrows coloured by the judgement instead of the fast/slow text.
Not a fan of the "flawless" though - empty text may work better, but would have to see how it feels during gameplay.

In general I think having two text lines is a bit cluttered.

File: osu.Game/Rulesets/Scoring/HitDetail.cs (hunk 6 above)
Comment (LevFlavien) [reply osu.Game/Rulesets/Scoring/HitDetail.cs:6]: How about `HitTiming` or `HitOffset`?

**Comment by skielred** [comment]
I updated my branch with current master, some stuffs that changed on master actually allowed rewriting some parts a bit cleaner.
This way if anyone want to still test or change things on this, at least it's up to date.

But feature-wise it's exactly the same as before and still suffer from the same problem as before.
The added fast/slow is still arbitrary put 20px above the judge, making it not very good visually using some skins as reported by https://github.com/ppy/osu/pull/7310#issuecomment-568040411 which is the top issue currently imo.

**Comment by mcendu** [comment]
Is this still being worked on?

**Comment by bdach** [comment]
given that the last commit was 11 months ago, i think the answer is "no". there is no point in bumping.

the most important part of this feature is going to be the visual design of the fast/slow display anyhow. the implementation is otherwise trivial. as i understand it this feature may come with future updates to the lazer default skins.

**Comment by peppy** [comment]
A visual design for this has already been decided, although I'm not sure if it's exposed yet on figma. @arflyte can you advise?

What I can say, is that this PR won't be satisfactory for implementing the new design, as it requires more than three states. Also there's enough conflicts in here to make a rewrite easier, so let's close for now.
